### DISCUSSION TITLE
architecture/declarative-api

### DISCUSSION THREAD

**Discussion by giscus** [isno/theByteBook#239] [General]
# architecture/declarative-api

构建大规模高可用的分布式系统

https://www.thebyte.com.cn/architecture/declarative-api.html

**Comment by xiaoronglv** [comment]
我的一点浅薄的观点，SQL 和“declarative” 没有太多关系，这个例子不太恰当。SQL 对立面的是 relation algorithm。


SQL: 

1. Provides a high-level interface for querying and updating data. 面向结果。


Relational Algorithm:

1. Refers to the theoretical, step-by-step procedures or processes used to perform operations on relational data. 面向过程

**Reply by isno** [reply]
wiki 百科中， SQL 定义为声明式语言。
https://zh.wikipedia.org/wiki/%E5%AE%A3%E5%91%8A%E5%BC%8F%E7%B7%A8%E7%A8%8B

SQL 属于典型的“明确目标，而非流程”的语言，和“declarative”（声明式） 有明确的关系。

**Reply by xiaoronglv** [reply]
我对 声明式 (Declarative) 和  Imperative 的理解，可能仅限于 Terraform / Kubernetes 的设计思想：

1. 定义一个 desired state
2. Terraform/Kubernetes 系统会把 current state 调整到 desired state

没想到 Declarative 居然有更广泛的含义，甚至 SQL 也属于 declarative ，涨知识了。

多谢你的解答。
